### ISSUE TITLE
Possibility to declare an optional field with a fallback value?

### ISSUE THREAD

**Issue by mathieul** [jamesmacaulay/elm-graphql#3]
Thank you for sharing this package and all the work you put into it!

This might be more a question or feature request than an issue, so please let me know if there is a better place to discuss it.

I'm failing to figure out how to declare a field with a hard-coded value which is missing in the GraphQL response. Thinking about something similar to [`Json.Decode.Pipeline.optional`](http://package.elm-lang.org/packages/NoRedInk/elm-decode-pipeline/3.0.0/Json-Decode-Pipeline#optional).

For instance if we had:

```elm
type alias Photo =
    { url : String
    , caption : String
    , editing : False
    }
```

and wanted to decode:

```graphql
query ($userID: ID!) {
  user(id: $userID) {
    name
    photos {
      url
      caption
    }
  }
}
```

so that `Photo.editing` is always `False`.

Thank you,
Mathieu

**Comment by jamesmacaulay** [comment]
Thanks for the feedback! I had functions to do more things like this in pre-release versions, but decided to take them out for 1.0.0. That decision was based on the idea that it would be easier to add functions later than to take them away, and that there's value in constraining the options provided for decoding to arbitrary types: it keeps the query definitions easier to read, the API surface of the library smaller, and it encourages developers to define types specific to the GraphQL responses, which I think is the right thing to do in most cases.

I likely will add back in an "escape hatch" or two to bypass this pattern, but I want to be thoughtful about how it's added and why.

It looks like your use case is that the `editing` field is local state relevant to the view that you want to associate with persistent remote state from the server. If that's the case, I would suggest keeping that `editing` flag separate, perhaps in a distinct record type that wraps the type you decode from the response. It kind of depends on how your model is structured, but it could look like this:

```elm
type alias Photo =
    { url : String
    , caption : String
    }

type alias PhotoEditor =
    { editing : Boolean
    , photo : Photo
    }
```

Then `PhotoEditor` is more explicitly a "view model" and it's more clear where each piece of data comes from and what it's concerned with.

I may be way off base with your requirements (please let me know) but this is the kind of approach I had in mind when choosing what to expose with the library. I'd love to get your thoughts on this.

**Comment by jamesmacaulay** [comment]
Also: the title of the issue and your reference to the `optional` function made me think you wanted a way of giving a field that might be null a default value, but your description of the use case made me think that you were looking for something more like [`hardcoded`](http://package.elm-lang.org/packages/NoRedInk/elm-decode-pipeline/3.0.0/Json-Decode-Pipeline#hardcoded). I can see the usefulness of both of those functions – are you looking for equivalents for both of them?

**Comment by mathieul** [comment]
While using the pre-1.0 release of `elm-graphql` I was using `produce` to achieve that (`|> withField "id" [] (produce False)`). I was indeed thinking about something like `hardcoded`. But indeed mixing model data with application state seems wrong now with likely coupling issues as my app grows. Your solution of wrapping items in an editor type looks cleaner and more flexible.

Thinking more about it, I started the application using strings for GraphQL queries and building my own JSON decoders for decoding responses. So I had the choice of mixing application state and data within the same type or going with the solution you recommend. I picked mixing state and data mostly because Elm doesn't have a succinct syntax for updating nested data (meaning you have to destructure the nested data first in order to update the photo data within the editor). I made that choice because I could, and it ends up being the wrong choice. If I had started with your library instead, I wouldn't have had that choice, so I would have had to go with the better type structure. So I can definitely see how constraining the options available can be a helpful tool to guide developers.

I can still think as some use cases that might legitimate such an option though. For instance if a todo list returns `TodoItem`s with a `completedOn` date field, we might want to convert the field to a `completed` boolean field. We'd want to do that not to have the API dictate the structure of the application when we don't have a say on the API itself. But that is an edge case, and I definitely see the advantage on tweaking the current API until you feel very confident about it before adding escape hatches to cover edge cases.

For now I will update my application following your advice and will let you know in the future if I run into a case where I think `hardcoded` or `optional` would have a more legitimate use.

Thanks again!

**Comment by jamesmacaulay** [comment]
I'm glad to hear that this approach will work for you, for now at least. I think it's likely that I'll add `produce` back into the library at some point, maybe with a note in its documentation that steers people away from it when it might not be a good idea. Another catch-all function that lets you transform the decoder of a spec however you like might also be a good idea, if its just a little bit painful to use then the library can still be opinionated about what the happy-path is, while letting people bypass that if they choose to. E.g. a `transformDecoder` function that you could use to implement your own `optional` equivalent and other things:

```elm
-- provided by elm-graphql:
transformDecoder :
    (Json.Decode.Decoder a -> Json.Decode.Decoder b)
    -> ValueSpec nullability coreType a vars
    -> ValueSpec nullability coreType b vars

-- provided by you:
mapResult :
    (a -> b)
    -> ValueSpec nullability coreType a vars
    -> ValueSpec nullability coreType b vars
mapResult f spec =
    transformDecoder (Json.Decode.map f) spec

nullableWithDefault :
    a
    -> ValueSpec NonNull coreType a vars
    -> ValueSpec Nullable coreType a vars
nullableWithDefault default spec =
    mapResult (Maybe.withDefault default) (nullable spec)
```

**Comment by jamesmacaulay** [comment]
Closing this issue in favour of #21 to make it clear what the proposed feature is. Meanwhile the `nullableWithDefault` function is now possible to write thanks to the newly-exposed `map` function:

http://package.elm-lang.org/packages/jamesmacaulay/elm-graphql/1.4.0/GraphQL-Request-Builder#map

**Comment by jamesmacaulay** [comment]
Implemented with the function `withLocalConstant` in 906e5c5dc4833616ac3b80ee1231567f303286c4, published in [1.7.0](http://package.elm-lang.org/packages/jamesmacaulay/elm-graphql/1.7.0/GraphQL-Request-Builder#withLocalConstant).
